scripts/standardising.py

```diff
@@ -189,5 +194,14 @@ def standardising(
         # Need GDAL to write to temporary location so no broken files end up in the done folder.
         run_gdal(command, input_file=input_file, output_file=standardized_working_path)
 
+        if tags(standardized_working_path)["TileByteCounts"].value[0] == 0:
```

Comment:
Is this just checking the first tile size? what about the rest of the tiles there could be 1000s?

Reply:
`exiftool` was displaying the single tile byte count as just "0", so when tifffile returned `(0,)` I thought that was just a quirk. But the problem was with `exiftool` displaying a single-entry list as a string value. You're right, we need to check that all the values are zero.

Reply:
What image in the COG this checking ? the root image (id:0) ? 

tile counts can be calculated with the tags `ceil(ImageWidth/TileWidth) * ceil(ImageHeight/TileHeight)`

Reply:
> What image in the COG this checking ? the root image (id:0) ?

[Yes](https://github.com/linz/topo-imagery/pull/767/commits/a06a5f27cbba1ccc3de81536d8a7044682d1277e#diff-8875e1bb3281897723637f0efc9c224ca12b38f42cd8a02419d86f632d36e1b4R207) (well, it's the first "page"). Do we need to check all pages?

> tile counts can be calculated with the tags `ceil(ImageWidth/TileWidth) * ceil(ImageHeight/TileHeight)`

I shouldn't need to do that though, right? Isn't checking the `TileByteCounts` tags enough?

Reply:
> I shouldn't need to do that though, right? Isn't checking the TileByteCounts tags enough?

I don't think so, but it could be interesting to validate the array sizes. 

I am not 100% sure what tifffile would do if there is only one tile, would it return `[0]` or just `0` ? 

> [Yes](https://github.com/linz/topo-imagery/pull/767/commits/a06a5f27cbba1ccc3de81536d8a7044682d1277e#diff-8875e1bb3281897723637f0efc9c224ca12b38f42cd8a02419d86f632d36e1b4R207) (well, it's the first "page"). Do we need to check all pages?

Nope, if the root is empty then everything is empty.

It can be done in reverse for performance reasons, querying the max overview is only 1 offset to fetch vs fetching 1000s (or even millions) of offsets for the root image.

Reply:
> I am not 100% sure what tifffile would do if there is only one tile, would it return `[0]` or just `0` ?

It returns `(0,)`, as mentioned.

> > [Yes](https://github.com/linz/topo-imagery/pull/767/commits/a06a5f27cbba1ccc3de81536d8a7044682d1277e#diff-8875e1bb3281897723637f0efc9c224ca12b38f42cd8a02419d86f632d36e1b4R207) (well, it's the first "page"). Do we need to check all pages?
> 
> Nope, if the root is empty then everything is empty.
> 
> It can be done in reverse for performance reasons, querying the max overview is only 1 offset to fetch vs fetching 1000s (or even millions) of offsets for the root image.

I don't understand the terminology. What's the "max overview"? Which "offsets"? Are these tags/other metadata/image properties of some other sort?

Reply:
> I don't understand the terminology. What's the "max overview"? Which "offsets"? Are these tags/other metadata/image properties of some other sort?

a COG is a collection of images, the root image and then a bunch of overviews or thumbnails, every overview halves the size of the image and quarters the number of tiles required

for a smallish LERC cog here are the overviews

```
        Id      Size            Tile Size       Tile Count      Resolution      Tile Stats                         
        0       32768x32768     512x512         64x64 (4096)    305.7481        684.18 MB (avg: 171 KB)  
        1       16384x16384     512x512         32x32 (1024)    611.4962        103.78 MB (avg: 104 KB)  
        2       8192x8192       512x512         16x16 (256)     1222.9924       32.05 MB (avg: 128 KB)   
        3       4096x4096       512x512         8x8 (64)        2445.9849       9.48 MB (avg: 152 KB)    
        4       2048x2048       512x512         4x4 (16)        4891.9698       2.69 MB (avg: 173 KB)    
        5       1024x1024       512x512         2x2 (4)         9783.9396       761 KB (avg: 190 KB)     
        6       512x512         512x512         1x1 (1)         19567.8792      205 KB (avg: 205 KB)     
```

if you fetch imageId:6 and look at the `TileOffsetBytes` there is only one record, for imageId:0  there are 4096 elements in the `TileOffsetBytes` array.

This is a somewhat small (dimension wise, its stil ~800mb) file some large tiffs will have 10's of thousands of tiles in imageId:0 but will generally only have 1 in `imageId[imageId.length - 1]` 

If you query the max image and it has data then there must be data in the tiff, for efficiency querying the top element, then only if that is empty do you query the root image.

In this specific case as you have direct access to the raw tiff and the tiff is generally small querying the root image's`TileOffsetBytes` is fine.


> It returns (0,), as mentioned. So it's always a tuple, and we can always iterate.

I am assuming this means you always iterate to ensure all values are 0.

Reply:
> This is a somewhat small (dimension wise, its stil ~800mb) file some large tiffs will have 10's of thousands of tiles in imageId:0 but will generally only have 1 in `imageId[imageId.length - 1]`

Cool, so I should just loop through the images in reverse order and write the file as soon as any of them are non-empty.

[The COG I'm testing on](https://s3.console.aws.amazon.com/s3/object/linz-topographic-upload?region=ap-southeast-2&prefix=aerialsurveys/CycloneGabrielle2023_PGRM3124/20230308_v1Orthos/tifs-RGB-v1/BF43_0205.tif) doesn't seem to have any `TileOffsetBytes` tags; did you mean `TileByteCounts` or is that property somewhere else?

Reply:
for your specific use case just use the root image, imageId:0

Reply:
Done.